src/internet/playmusic/playmusicservice.cpp

```diff
@@ -0,0 +1,562 @@
+const char* PlayMusicService::kAuthUrl = "https://www.google.com/accounts/ClientLogin";
+
+const int PlayMusicService::kSearchDelayMsec = 400;
+const int PlayMusicService::kSongSearchLimit = 100;
+
+typedef QPair<QString, QString> Param;
+
+PlayMusicService::PlayMusicService(Application* app, InternetModel* parent)
+        : InternetService(kServiceName, app, parent, parent),
+          root_(nullptr),
+          search_(nullptr),
+          network_(new NetworkAccessManager(this)),
+          context_menu_(nullptr),
+          search_box_(new SearchBoxWidget(this)),
+          search_delay_(new QTimer(this)),
+          next_pending_search_id_(0) {
+  qsrand(QTime::currentTime().msec());
+
+  QByteArray s1 = QByteArray::fromBase64(kS1);
+  QByteArray s2 = QByteArray::fromBase64(kS2);
+  QByteArray key = QByteArray(s1.size(), '\0');
+  for (int i = 0; i < s1.length(); i++) {
+    key[i] = s1[i]^s2[i];
+  }
+  key_ = QString(key);
+
+  search_delay_->setInterval(kSearchDelayMsec);
+  search_delay_->setSingleShot(true);
+  connect(search_delay_, SIGNAL(timeout()), SLOT(DoSearch()));
+
+  PlayMusicSearchProvider* search_provider =
+          new PlayMusicSearchProvider(app_, this);
+  search_provider->Init(this);
+  app_->global_search()->AddProvider(search_provider);
+
+  connect(search_box_, SIGNAL(TextChanged(QString)), SLOT(Search(QString)));
+  app->player()->RegisterUrlHandler(new PlayMusicUrlHandler(this, this));
+}
+
+PlayMusicService::~PlayMusicService() {}
+
+QStandardItem* PlayMusicService::CreateRootItem() {
+  root_ = new QStandardItem(QIcon(":providers/playmusic.png"), kServiceName);
+  root_->setData(true, InternetModel::Role_CanLazyLoad);
+  root_->setData(InternetModel::PlayBehaviour_DoubleClickAction,
+                 InternetModel::Role_PlayBehaviour);
+  return root_;
+}
+
+void PlayMusicService::LazyPopulate(QStandardItem* item) {
+  switch (item->data(InternetModel::Role_Type).toInt()) {
+    case InternetModel::Type_Service: {
+      EnsureItemsCreated();
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+void PlayMusicService::EnsureItemsCreated() {
+  if (!search_) {
+    search_ =
+            new QStandardItem(IconLoader::Load("edit-find"), tr("Search results"));
+    search_->setToolTip(
+            tr("Start typing something on the search box above to "
+                       "fill this search results list"));
+    search_->setData(InternetModel::PlayBehaviour_MultipleItems,
+                     InternetModel::Role_PlayBehaviour);
+    root_->appendRow(search_);
+  }
+}
+
+QWidget* PlayMusicService::HeaderWidget() const { return search_box_; }
+
+void PlayMusicService::ShowConfig() {
+  app_->OpenSettingsDialogAtPage(SettingsDialog::Page_PlayMusic);
+}
+
+void PlayMusicService::Homepage() {
+  QDesktopServices::openUrl(QUrl("https://play.google.com/"));
+}
+
+void PlayMusicService::Login(const QString &username, const QString &password) {
```

Comment:
Not thrilled about this - presumably this access token does not work for other services?

And I guess this will fail if 2-step verification is enabled